js-watcher/cache.go

```diff
@@ -0,0 +1,150 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"log"
+	"path/filepath"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/rest"
+	k8sCache "k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
+)
+
+// Cache is a light wrapper around a k8s informer for a single k8s namespace.
+type Cache struct {
+	lister      k8sCache.GenericLister
+	informer    k8sCache.SharedInformer
+	namespace   string
+	userIDLabel string
+}
+
+// GenericResource allows unmarshalling the metadata of dynamic resources without
+// unmarshalling their status and spec.
+type GenericKubernetesResource struct {
+	Metadata metav1.ObjectMeta `json:"metadata"`
+	Spec     json.RawMessage   `json:"spec"`
+	Status   json.RawMessage   `json:"status,omitempty"`
+
+	metav1.TypeMeta `json:",inline"`
+}
+
+// synchronize synchronizes the the informer cache. It uses a channel that it will
+// write to a boolean value to indicate whether the synchronization completed successfully
+// (true) or it did not complete (false).
+func (c *Cache) synchronize(ctx context.Context, doneCh chan bool) {
+	if !k8sCache.WaitForCacheSync(ctx.Done(), c.informer.HasSynced) {
+		log.Print("Could not sync informer cache with k8s.")
+		doneCh <- false
+	}
+	doneCh <- true
+}
+
+// run runs the informer.
+func (c *Cache) run(ctx context.Context) {
+	c.informer.Run(ctx.Done())
+}
+
+func (c *Cache) getAll() (res []runtime.Object, err error) {
+	res, err = c.lister.List(labels.NewSelector())
+	if err != nil {
+		return nil, fmt.Errorf("could not list servers for namespace %s: %w", c.namespace, err)
+	}
+	return res, nil
+}
+
+// getByUserID retrieves resources that belong to a specific user from the informer cache.
+func (c *Cache) getByUserID(userID string) (res []runtime.Object, err error) {
+	selector := labels.NewSelector()
+	if userID != "" {
+		requirement, err := labels.NewRequirement(c.userIDLabel, selection.Equals, []string{userID})
+		if err != nil {
+			return nil, fmt.Errorf("could not set up selector when looking for servers for userID %s: %w", userID, err)
+		}
+		selector = selector.Add(*requirement)
+	}
+	res, err = c.lister.List(selector)
+	if err != nil {
+		return nil, fmt.Errorf("could not list servers for userID %s: %w", userID, err)
+	}
+	return res, nil
+}
+
+// getByName retrieves a specific resource from the informer cache.
+func (c *Cache) getByName(name string) (res runtime.Object, err error) {
+	res, err = c.lister.Get(fmt.Sprintf("%s/%s", c.namespace, name))
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
```

Comment:
Not sure if you're deliberately trying to be explicit here but the return vals do not need to be specified here; in fact, I think you can leave out the `return` completely.

Reply:
Oh cool I did not know that. It cannot be used in all cases. But in most I can and I did so.

It feels wrong coming from python. Because that always results in None being returned in python.